File: src/main/kotlin/lec_17/Main17.java
package lec_17;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Main17 {
  public static void main(String[] args) {
    List<Fruit> fruits = Arrays.asList(
      new Fruit("사과", 1000),
      new Fruit("사과", 1200),
      new Fruit("사과", 1200),
      new Fruit("사과", 1500),
      new Fruit("바나나", 3000),
      new Fruit("바나나", 3200),
      new Fruit("바나나", 2500),
      new Fruit("수박", 10000)
    );

    // 익명클래스를 작성하여 변수로 넘긴다. 그러나 이 방법은 복잡하며, 다양한 filter가 필요할 수 있다.
    filterFruits(fruits, new FruitFilter() {
      @Override
      public boolean isSelected(Fruit fruit) {
        return Arrays.asList("사과", "바나나").contains(fruit.getName()) && fruit.getPrice() > 5000;
      }
    });

    // 람다는 깔끔하고 간결하게 사용 가능하다.
    filterFruits(fruits, fruit -> fruit.getName().equals("사과"));

    /**
     * java는 함수를 변수 및 파라미터로 취급할 수 없는 2급 시민이다.
     * 따라서 함수를 넘겨주고 있는 것처럼 보이지만 사실은 그렇지 않다.
      */
    filterFruitsWithLambda(fruits, Fruit::isApple);
  }

  // 사과만 찾고 싶을 때
  private List<Fruit> findApples(List<Fruit> fruits) {
    List<Fruit> apples = new ArrayList<>();
    for (Fruit fruit : fruits) {
      if (fruit.getName().equals("사과")) apples.add(fruit);
    }

    return apples;
  }

  // 바나나만 찾고 싶을 때
  private List<Fruit> findBananas(List<Fruit> fruits) {
    List<Fruit> bananas = new ArrayList<>();
    for (Fruit fruit : fruits) {
      if (fruit.getName().equals("바나나")) bananas.add(fruit);
    }

    return bananas;
  }

  // 위 두 개 메서드는 중복이 발생하므로 리팩터링
  private List<Fruit> findFruitsWithName(List<Fruit> fruits, String name) {
    List<Fruit> results = new ArrayList<>();
    for (Fruit fruit : fruits) {
      if (fruit.getName().equals(name)) results.add(fruit);
    }

    return results;
  }

  /**
   * 사과랑 바나나 같이 보여달라
   * 1200원 넘지 않는 사과만 보여달라 등등..
   * 조건이 늘어나면 늘어날수록 메서드도 늘어나게 된다.
   * 따라서 조건을 추상화하여 변수로 받아 사용한다.
   */
  private static List<Fruit> filterFruits(List<Fruit> fruits, FruitFilter fruitFilter) {
    List<Fruit> results = new ArrayList<>();
    for (Fruit fruit : fruits) {
      if (fruitFilter.isSelected(fruit)) results.add(fruit);
    }

    return results;
  }

  // 람다를 사용하여 변경
  private static List<Fruit> filterFruitsWithLambda(List<Fruit> fruits, Predicate<Fruit> fruitFilter) {

    return fruits.stream()
      .filter(fruitFilter)
      .collect(Collectors.toList());
  }
}

File: src/main/kotlin/lec_17/Main17.kt
package lec_17

fun main() {
    val fruits = listOf(
        Fruit("사과", 1000),
        Fruit("사과", 1200),
        Fruit("사과", 1200),
        Fruit("사과", 1500),
        Fruit("바나나", 3000),
        Fruit("바나나", 3200),
        Fruit("바나나", 2500),
        Fruit("수박", 10000),
    )

    // 함수 이름없이 바로 파라미터 / 리턴타입 / 본문이 들어감.
    val isApple = fun(fruit: Fruit): Boolean {
        return fruit.name == "사과"
    }

    // 람다 방식 2
    val isApple2 = { fruit: Fruit -> fruit.name == "사과" }

    // 람다를 직접 호출하는 방법
    isApple(fruits[0])

    // invoke를 사용하여 명시적으로 작성할수도 있음.
    isApple.invoke(fruits[0])


    // 함수에 타입을 지정할 수 있다. (Fruit를 받고, Boolean을 반환)
    val isApple3: (Fruit) -> Boolean = fun(fruit: Fruit): Boolean {
        return fruit.name == "사과"
    }

    val isApple4: (Fruit) -> Boolean = { fruit: Fruit -> fruit.name == "사과" }

    // Fruit를 변수로 받고, Boolean을 반환하는 함수를 타입으로 받기 때문에 함수를 매개변수로 넘길 수 있다.
    filterFruits(fruits, isApple)
    filterFruits(fruits, isApple2)

    // 소괄호 안에 중괄호가 들어가 있는 것이 어색하기 때문에 밖으로 뺄 수 있다.
    filterFruits(fruits, { fruit: Fruit -> fruit.name == "사과" })

    // 밖으로 뺀 형태
    filterFruits(fruits) { fruit: Fruit -> fruit.name == "사과" }

    // 타입을 추론 가능하기 때문에 생략 가능
    filterFruits(fruits) { fruit -> fruit.name == "사과" }

    // it을 사용하면 람다의 파라미터를 직접 참조할 수 있다. 그러나 이 방법은 어떤 값이 넘어왔는지 알 수 없으므로 비추천.
    filterFruits(fruits) { it.name == "사과" }


    /**
     * java에서는 final이 아닌 변수는 람다에서 사용할 수 없으나,
     * kotlin에서는 람다가 시작하는 지점에 참조하고 있는 변수들을 모두 포획하여 그 정보를 가지고 있다.
     * 따라서, 바나나 -> 수박으로 바뀌어도 아무 문제 없이 사용할 수 있다.
     * 이렇게 해야만 람다를 진정한 1급 시민으로 간주할 수 있으며, 이러한 데이터 구조를 Closure라고 부른다.
     */
    var targetFruitName = "바나나"
    targetFruitName = "수박"
    filterFruits(fruits) { it.name == targetFruitName }
}

private fun filterFruits(
    fruits: List<Fruit>, filter: (Fruit) -> Boolean
): List<Fruit> {
    val results = mutableListOf<Fruit>()
    for (fruit in fruits) {
        if (filter(fruit)) results.add(fruit)
    }

    return results
}